File: Lab3/src/main/java/ubb/scs/map/repository/file/AbstractFileRepository.java
package ubb.scs.map.repository.file;

import ubb.scs.map.domain.Entity;
import ubb.scs.map.domain.validators.Validator;
import ubb.scs.map.repository.memory.InMemoryRepository;

import java.io.*;

public abstract class AbstractFileRepository<ID, E extends Entity<ID>> extends InMemoryRepository<ID, E> {
    private final String filename;

    public AbstractFileRepository(Validator<E> validator, String fileName) {
        super(validator);
        filename = fileName;
        readFromFile();
    }


    public abstract E createEntity(String line);

    public abstract String saveEntity(E entity);

    @Override
    public E save(E entity) {
        E e = super.save(entity);
        if (e == null) writeToFile();
        return e;
    }

    private void writeToFile() {

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (E entity : entities.values()) {
                String ent = saveEntity(entity);
                writer.write(ent);
                writer.newLine();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

    private void readFromFile() {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                E entity = createEntity(line);
                super.save(entity);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public E delete(ID id) {
        E entity = super.delete(id);
        writeToFile(); //nu i null
        return entity;
    }

    @Override
    public E update(E entity) {
        E _entity = super.update(entity);
        writeToFile(); // la fel
        return _entity;
    }
}

File: Lab10/Lab10/Repository/AbstractFileRepository.cs
using System;
using System.Collections.Generic;
using System.IO;
using Lab10.Domain.Validators;

namespace Lab10.Domain;

public abstract class AbstractFileRepository<ID, E> : InMemoryRepository<ID, E> where E : Entity<ID>
{
    private readonly string filename;

    public AbstractFileRepository(IValidator<E> validator, string fileName) : base(validator)
    {
        filename = fileName;
    }

    protected void Initialize()
    {
        ReadFromFile();
    }

    public abstract E CreateEntity(string line);

    public abstract string SaveEntity(E entity);

    public override E Save(E entity)
    {
        var e = base.Save(entity);
        if (e == null) WriteToFile();
        return e;
    }

    private void WriteToFile()
    {
        try
        {
            using (var writer = new StreamWriter(filename))
            {
                foreach (var entity in entities.Values)
                {
                    var ent = SaveEntity(entity);
                    writer.WriteLine(ent);
                }
            }
        }
        catch (IOException e)
        {
            throw new Exception("Error writing to file", e);
        }
    }

    private void ReadFromFile()
    {
        try
        {
            using (var reader = new StreamReader(filename))
            {
                string line;
                while ((line = reader.ReadLine()) != null)
                {
                    var entity = CreateEntity(line);
                    base.Save(entity);
                }
            }
        }
        catch (IOException e)
        {
            throw new Exception("Error reading from file", e);
        }
    }

    public override E Delete(ID id)
    {
        var e = base.Delete(id);
        WriteToFile();
        return e;
    }

    public override E Update(E entity)
    {
        var e = base.Update(entity);
        WriteToFile();
        return e;
    }
}